src/core/core.controller.js

```diff
@@ -969,19 +969,24 @@ helpers.extend(Chart.prototype, /** @lends Chart */ {
 		});
 	},
 
-	updateHoverStyle: function(elements, mode, enabled) {
+	updateHoverStyle: function(items, mode, enabled) {
 		var prefix = enabled ? 'set' : 'remove';
-		var element, i, ilen;
+		var meta, item, i, ilen;
 
-		for (i = 0, ilen = elements.length; i < ilen; ++i) {
-			element = elements[i];
-			if (element) {
-				this.getDatasetMeta(element._datasetIndex).controller[prefix + 'HoverStyle'](element);
+		if (mode === 'dataset') {
+			meta = this.getDatasetMeta(items[0].datasetIndex);
+			meta.controller['_' + prefix + 'DatasetHoverStyle']();
+			for (i = 0, ilen = meta.data.length; i < ilen; ++i) {
```

Comment:
If I understand correctly, then I think this is the same way the code currently works:
https://github.com/chartjs/Chart.js/blob/a2bf51a73c244f11204b972dcd1cb0d2b6750810/src/core/core.interaction.js#L175

I did a bit of an optimization here though. When it's dataset I pass only the `datasetIndex` because that was enough information to reproduce the current functionality and I didn't want to create a new `item` for every data point.

Reply:
And when the interaction gets fixed, this would need fixing too. Should use the items provided, not replicate the functionality.

Reply:
I didn't actually replicate the functionality, but just moved where it occurs as a performance optimization. I can move it back if we think it's over optimizing

Reply:
Let me have another look, I might have missed something.